File: solutions/go/strain/1/strain.go
package strain

func Keep[T any](s []T, f func(T) bool) []T {
	var out []T
	for _, v := range s {
		if f(v) {
			out = append(out, v)
		}
	}
	return out
}

func Discard[T any](s []T, f func(T) bool) []T {
	var out []T
	for _, v := range s {
		if !f(v) {
			out = append(out, v)
		}
	}
	return out
}

File: solutions/csharp/strain/1/Strain.cs
public static class Strain
{
    public static IEnumerable<T> Keep<T>(this IEnumerable<T> collection, Func<T, bool> predicate) => Apply(collection, predicate, true);

    public static IEnumerable<T> Discard<T>(this IEnumerable<T> collection, Func<T, bool> predicate) => Apply(collection, predicate, false);

    public static IEnumerable<T> Apply<T>(this IEnumerable<T> collection, Func<T, bool> predicate, bool result)
    {
        for (var i = 0; i < collection.Count(); i++)
            if (predicate(collection.ElementAt(i)) == result)
                yield return collection.ElementAt(i);
    }
}